bokehjs/src/coffee/models/plots/plot_canvas.coffee

```diff
@@ -470,11 +467,33 @@ class PlotCanvasView extends Renderer.View
 
     ctx.restore()  # Restore to default state
 
+    # Invoke a resize on the document the first time that all plots of that
+    # document are rendered. For some reason, the layout solver only works well
+    # after the plots have been rendered. See #4401.
+    if @model.document._unrendered_plots?
+      delete @model.document._unrendered_plots[@id]
+      if _.isEmpty(@model.document._unrendered_plots)
+        @model.document._unrendered_plots = null
+        _.delay($.proxy(@model.document.resize, @model.document), 1)
+
   resize: () ->
+    # Set the plot and canvas to the current model's size
+    # This gets called upon solver resize events
     width = @model._width._value
     height = @model._height._value
 
-    @canvas_view.set_dims([width, height], true)
+    @canvas_view.set_dims([width, height], true)  # this indirectly calls @request_render
+
+    # Prepare the canvas size, taking HIDPI into account. Note that this may cause
+    # a resize of the canvas, which means that any previous calls to ctx.save() may be undone.
+    @canvas_view.prepare_canvas()
+
+    try
+      @update_constraints()
+    catch silent_error
```

Comment:
I have long wished that kiwi.js threw errors up more gracefully so that we could handle them cleanly. @bryevdv and I had been talking about making some improvements to kiwi.js - this could be one of them.